src/interpreter/executors/predicate.rs

```diff
@@ -43,4 +45,338 @@ where
+                self.gas_charge(GAS_EXP)?;
+                self.alu_boolean_overflow(ra, Word::overflowing_pow, b, c as u32)?;
+            }
+
+            OpcodeRepr::EXPI => {
+                self.gas_charge(GAS_EXPI)?;
+                self.alu_boolean_overflow(ra, Word::overflowing_pow, b, imm as u32)?;
+            }
+
+            OpcodeRepr::GT => {
+                self.gas_charge(GAS_GT)?;
+                self.alu_set(ra, (b > c) as Word)?;
+            }
+
+            OpcodeRepr::LT => {
+                self.gas_charge(GAS_LT)?;
+                self.alu_set(ra, (b < c) as Word)?;
+            }
+
+            OpcodeRepr::MLOG => {
+                self.gas_charge(GAS_MLOG)?;
+                self.alu_error(
+                    ra,
+                    |b, c| (b as f64).log(c as f64).trunc() as Word,
+                    b,
+                    c,
+                    b == 0 || c <= 1,
+                )?;
+            }
+
+            OpcodeRepr::MOD => {
+                self.gas_charge(GAS_MOD)?;
+                self.alu_error(ra, Word::wrapping_rem, b, c, c == 0)?;
+            }
+
+            OpcodeRepr::MODI => {
+                self.gas_charge(GAS_MODI)?;
+                self.alu_error(ra, Word::wrapping_rem, b, imm, imm == 0)?;
+            }
+
+            OpcodeRepr::MOVE => {
+                self.gas_charge(GAS_MOVE)?;
+                self.alu_set(ra, b)?;
+            }
+
+            OpcodeRepr::MOVI => {
+                self.gas_charge(GAS_MOVI)?;
+                self.alu_set(ra, imm)?;
+            }
+
+            OpcodeRepr::MROO => {
+                self.gas_charge(GAS_MROO)?;
+                self.alu_error(
+                    ra,
+                    |b, c| (b as f64).powf((c as f64).recip()).trunc() as Word,
+                    b,
+                    c,
+                    c == 0,
+                )?;
+            }
+
+            OpcodeRepr::MUL => {
+                self.gas_charge(GAS_MUL)?;
+                self.alu_capture_overflow(ra, u128::overflowing_mul, b.into(), c.into())?;
+            }
+
+            OpcodeRepr::MULI => {
+                self.gas_charge(GAS_MULI)?;
+                self.alu_capture_overflow(ra, u128::overflowing_mul, b.into(), imm.into())?;
+            }
+
+            OpcodeRepr::NOOP => {
+                self.gas_charge(GAS_NOOP)?;
+                self.alu_clear()?;
+            }
+
+            OpcodeRepr::NOT => {
+                self.gas_charge(GAS_NOT)?;
+                self.alu_set(ra, !b)?;
+            }
+
+            OpcodeRepr::OR => {
+                self.gas_charge(GAS_OR)?;
+                self.alu_set(ra, b | c)?;
+            }
+
+            OpcodeRepr::ORI => {
+                self.gas_charge(GAS_ORI)?;
+                self.alu_set(ra, b | imm)?;
+            }
+
+            OpcodeRepr::SLL => {
+                self.gas_charge(GAS_SLL)?;
+                self.alu_set(ra, b.checked_shl(c as u32).unwrap_or_default())?;
```

Comment:
unfortunate that we have to copy the instruction execution implementation 😕 

This creates a possibility for a drift in behavior that could be hard to detect or test for.

Reply:
Agree, but this change is more delicate since we have different storage backends for these executions. Can we leave this for the next week as debt so I come with a more complete solution that won't duplicate code?

Reply:
yeah makes sense, we just need to address this before bit-rot sets in